sharkfin/population.py

```diff
@@ -97,94 +214,116 @@ def class_stats(self, store=False):
+        cs["aLvl_mean"] = cs["aLvl"]["mean"]
+        cs["aLvl_std"] = cs["aLvl"]["std"]
+        cs["mNrm_mean"] = cs["mNrm"]["mean"]
+        cs["mNrm_std"] = cs["mNrm"]["std"]
+        cs["mNrm_ratio_StE_mean"] = cs["mNrm_ratio_StE"]["mean"]
+        cs["mNrm_ratio_StE_std"] = cs["mNrm_ratio_StE"]["std"]
 
         if store:
             self.stored_class_stats = class_stats
 
         return class_stats
 
-    def create_distributed_agents(self, agent_parameters, dist_params, n_per_class):
-        """
-        Creates agents of the given classes with stable parameters.
-        Will overwrite the DiscFac with a distribution from CSTW_MPC.
+    def create_distributed_agents(self):
 
-        Parameters
-        ----------
-        agent_parameters: dict
-            Parameters shared by all agents (unless overwritten).
+        self.agents = [
+            self.agent_class.__class__(**agent_dict) for agent_dict in self.agent_dicts
+        ]
 
-        dist_params: dict of dicts
-            Parameters to distribute agents over, with discrete Uniform arguments
+    def create_database(self):
 
-        n_per_class: int
-            number of agents to instantiate per class
-        """
-        num_classes = math.prod([dist_params[dp]['n'] for dp in dist_params])
-        agent_batches = [{'AgentCount': num_classes}] * n_per_class
+        database = pd.DataFrame(self.agent_dicts)
+        database["agents"] = self.agents
 
-        agents = [
-            cpm.PortfolioConsumerType(
-                seed=random.randint(0, 2**31 - 1),
-                **update_return(agent_parameters, ac),
-            )
-            for ac in agent_batches
-        ]
+        self.database = database
 
-        agents = AgentList(agents, dist_params)
+    def solve_distributed_agents(self):
+        # see Market class for an example of how to solve distributed agents in parallel
 
-        return agents
+        for agent in self.agents:
+            agent.solve()
+
+    def unpack_solutions(self):
 
-    def init_simulation(self):
+        self.solution = [agent.solution for agent in self.agents]
+
+    def init_simulation(self, T_sim=1000):
         """
         Sets up the agents with their state for the state of the simulation
         """
         for agent in self.agents:
-            agent.track_vars += ['pLvl', 'mNrm', 'cNrm', 'Share', 'Risky']
+            agent.track_vars += ["pLvl", "mNrm", "cNrm", "Share", "Risky"]
+            agent.T_sim = T_sim
+            agent.initialize_sim()
 
-            agent.assign_parameters(AdjustPrb=1.0)
-            agent.T_sim = 1000  # arbitrary!
-            agent.solve()
 
-            agent.initialize_sim()
+class AgentPopulationSolution:
+    def __init__(self, agent_population):
+        self.agent_population = agent_population
+
+        self.dist_params = self.agent_population.dist_params
+        self.database = self.agent_population.database
+
+    def merge_solutions(self, continuous_states):
```

Comment:
@sbenthall need review and comments for this. 

I tried to start generic, but couldn't wrap my head around it, so I did it using the framework I know to expect...